R/get_zoltar_truth.R

```diff
@@ -0,0 +1,121 @@
+#' Default to NULL which will load the latest truth data from covidData. 
+#' 
+#' @return data frame with timezero, unit, target and value
+#'
+#' @export
+preprocess_truth_for_zoltar <- function(target, issue_date = NULL){
+  
+  # validate target 
+  target <- match.arg(target, 
+                      choices = c("Cumulative Deaths", "Incident Deaths"), 
+                      several.ok = FALSE)
+  
+  # load the most up to date weeky truth data from JHU CSSE
+  df <- covidData::load_jhu_data(issue_date = issue_date, 
+                                 spatial_resolution = c('national', 'state'),
+                                 temporal_resolution = 'weekly', 
+                                 measure = 'deaths',
+                                 replace_negatives = FALSE, 
+                                 adjustment_cases = 'none',
+                                 adjustment_method = 'none')
+  
+  # select columns needed for target
+  if (target == "Cumulative Deaths"){
+    target_var <- " wk ahead cum death"
+    df <- df %>%
+      dplyr::select(-inc) %>%
+      dplyr::rename(value = cum, unit = location)
+  } else {
+    target_var <- " wk ahead inc death"
+    df <- df %>%
+      dplyr::select(-cum) %>%
+      dplyr::rename(value = inc, unit = location)
+  }
+  
+  df$horizon <- 20
+  df <- df[rep(row.names(df), df$horizon),]
+  df <- df %>%
+    dplyr::group_by(unit, date, value) %>%
+    # going backward in time
+    dplyr::mutate(horizon = seq(-1,-20)) %>%
+    dplyr::ungroup() %>%
+    dplyr::mutate(
+      # generate the Saturday target end date in a number of weeks behind
+      target_end_date = calc_target_week_end_date(date, horizon),
+      # get the next Monday relative to Saturday target end date
+      timezero = lubridate::ceiling_date(as.Date(target_end_date), 'week') + 1,
+      horizon = abs(horizon),
+      target = paste0(horizon,target_var)) %>%
+    dplyr::select(unit, date, value, timezero, horizon,target)
+  
+  # set up Zoltar connection
+  zoltar_connection <- zoltr::new_connection()
+  if(Sys.getenv("Z_USERNAME") == "" | Sys.getenv("Z_PASSWORD") == "") {
+    zoltr::zoltar_authenticate(zoltar_connection, "zoltar_demo","Dq65&aP0nIlG")
+  } else {
+    zoltr::zoltar_authenticate(zoltar_connection, Sys.getenv("Z_USERNAME"),Sys.getenv("Z_PASSWORD"))
+  }
+  
+  # construct Zoltar project url
+  the_projects <- zoltr::projects(zoltar_connection)
+  project_url <- the_projects[the_projects$name == "COVID-19 Forecasts", "url"]
+  
+  # get all valid timezeros from zoltar
+  zoltar_timezeros<- zoltr::timezeros(zoltar_connection, project_url)$timezero_date
+  
+  
+  # generate target end date with horzion 1 to 20 for all zoltar_timezeros
+  df_zoltar <- data.frame(zoltar_timezeros = zoltar_timezeros, horizon = 20)
+  df_zoltar <- df_zoltar[rep(row.names(df_zoltar), df_zoltar$horizon),]
+  df_zoltar <- df_zoltar %>%
+    dplyr::group_by(zoltar_timezeros) %>%
+    dplyr::mutate(horizon = seq(1,20),
```

Comment:
I think this kind of operation could be more easily achieved and more readable by using

```
tidyr::expand_grid(
  zoltar_timezeros = zoltar_timezeros, horizon = seq_len(20)
)
```